File: leetcode_java/647.PalindromicSubstrings.java
// Time: O(n^2), Space: O(1)
class Solution {
    public int countSubstrings(String s) {
        int ans = 0;
        for (int i = 0; i < s.length(); ++i) {
            ans += countPalindromesAroundCenter(s, i, i);
            ans += countPalindromesAroundCenter(s, i, i + 1);
        }
        return ans;
    }

    private int countPalindromesAroundCenter(String s, int lo, int hi) {
        int ans = 0;
        for (; lo >= 0 && hi < s.length(); -- lo, ++ hi) {
            if (s.charAt(lo) != s.charAt(hi)) break;
            ++ ans;
        }
        return ans;
    }
}

/**
647. Palindromic Substrings
Medium

Given a string s, return the number of palindromic substrings in it.

A string is a palindrome when it reads the same backward as forward.

A substring is a contiguous sequence of characters within the string.


Example 1:

Input: s = "abc"
Output: 3
Explanation: Three palindromic strings: "a", "b", "c".

Example 2:

Input: s = "aaa"
Output: 6
Explanation: Six palindromic strings: "a", "a", "a", "aa", "aa", "aaa".
 */

File: leetcode/647.PalindromicSubstrings.cpp
// Time Complexity: O(N^2)
// Space: O(1)
class Solution {
public:
    int countSubstrings(string s) {
        int n = s.length(), count = 0;

        for (int i = 0; i < n; ++ i) {
            count += getCount(s, i, i) + getCount(s, i, i+1);
        }
        return count;
    }

    int getCount(string& s, int left, int right) {
        int count = 0;
        while (left >= 0 && right < s.length() && s[left] == s[right]) {
            ++ count;
            -- left, ++ right;
        };
        return count;
    }
};

/**
647. Palindromic Substrings
Medium

Given a string s, return the number of palindromic substrings in it.

A string is a palindrome when it reads the same backward as forward.

A substring is a contiguous sequence of characters within the string.
 */
